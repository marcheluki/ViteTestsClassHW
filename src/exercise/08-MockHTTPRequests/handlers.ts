import { rest } from 'msw'

export const handlers = [
  rest.post('https://auth-provider.example.com/api/login', async (req, res, ctx) => {
    const { username, password } = await req.json()
    
    if (!password) {
      return res(
        ctx.status(400),
        ctx.json({ message: 'Password is required' })
      )
    }

    if (!username) {
      return res(
        ctx.status(400),
        ctx.json({ message: 'Username is required' })
      )
    }

    return res(
      ctx.json({
        username
      })
    )
  })
]