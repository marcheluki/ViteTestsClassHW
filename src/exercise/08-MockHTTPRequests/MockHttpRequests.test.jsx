import React from 'react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import LoginSubmission from '../sharedComponent/LoginSubmission'
import { handlers } from './handlers'

const server = setupServer(...handlers)

// Setup
beforeAll(() => server.listen())
afterEach(() => {
  server.resetHandlers()
})
afterAll(() => server.close())

describe('LoginSubmission', () => {
  test('successful login submission', async () => {
    const user = userEvent.setup()
    render(<LoginSubmission />)

    const username = screen.getByLabelText(/username/i)
    const password = screen.getByLabelText(/password/i)
    const submitButton = screen.getByRole('button', { name: /submit/i })

    await user.type(username, 'testuser')
    await user.type(password, 'testpass')
    await user.click(submitButton)

    // Wait for spinner to be removed
    await waitForElementToBeRemoved(() => screen.getByLabelText('loading...'))

    // Assert success message
    expect(screen.getByText(/welcome/i)).toBeInTheDocument()
    expect(screen.getByText('testuser')).toBeInTheDocument()
  })

  test('handles server error', async () => {
    server.use(
      rest.post('https://auth-provider.example.com/api/login', (req, res, ctx) => {
        return res(
          ctx.status(500),
          ctx.json({ message: 'Internal server error' })
        )
      })
    )

    const user = userEvent.setup()
    render(<LoginSubmission />)

    const username = screen.getByLabelText(/username/i)
    const password = screen.getByLabelText(/password/i)
    
    await user.type(username, 'testuser')
    await user.type(password, 'testpass')
    await user.click(screen.getByRole('button', { name: /submit/i }))

    await waitForElementToBeRemoved(() => screen.getByLabelText('loading...'))
    
    expect(screen.getByRole('alert')).toMatchSnapshot()
  })

  test('handles missing password', async () => {
    const user = userEvent.setup();
    render(<LoginSubmission />);

    const username = screen.getByLabelText(/username/i);
    await user.type(username, 'testuser');
    await user.click(screen.getByRole('button', { name: /submit/i }));

    // Wait for spinner removal
    await waitForElementToBeRemoved(() => screen.getByLabelText('loading...'));
    
    // Assert error message is rendered.
    expect(screen.getByRole('alert')).toMatchSnapshot();
  });
})