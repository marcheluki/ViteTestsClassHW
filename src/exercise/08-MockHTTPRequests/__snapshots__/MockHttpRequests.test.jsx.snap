// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSubmission handles missing password 1`] = `
<div
  role="alert"
  style="color: red;"
>
  Password is required
</div>
`;

exports[`LoginSubmission handles server error 1`] = `
<div
  role="alert"
  style="color: red;"
>
  Internal server error
</div>
`;
